(feature): 게시물 생성 페이지 option 항목 게시물 생성 페이지 노출

Derek-94 opened the pull request:
### 구현 내용

- 게시물 목록 페이지 -> 글쓰기 페이지로 이동 후 옵션 처리.

  - **게시물 목록 페이지**에서 이동되어야 함.

- 해당 게시물 목록 페이지가 가지고 있는 옵션들 (분반 정보 및 질문 유형 정보들) 기반으로 글쓰기 페이지에서 그 옵션들 중 선택 가능함.

- 글쓰기 화면에서 옵션 선택시 제목위에 바로 렌더링 됨.

- 게시글을 작성 후 등록 시, `container` 화면에서 새롭게 만든 게시물 (옵션 포함) 정보 출력 확인.

---

### 상세내용

1. 먼저, `BoardDetailContainer` 가 `api` 통신을 통해 얻은 정보를 `atom` 으로 관리합니다.

![image](https://user-images.githubusercontent.com/52649378/122151691-f1167b80-ce9a-11eb-8f7c-c925992837e8.png)

그럼, 이제 위에 보이는 `articleCreateOption` atom 에는 아래 사진 처럼 정보가 있겠죠.

![image](https://user-images.githubusercontent.com/52649378/122151486-9c730080-ce9a-11eb-8046-1fc83b0b8c15.png)

**위 사진 처럼 `options` 정보가 들어가 있는 `atom` 을 기반으로 게시물 글쓰기 페이지에서 저 정보를 렌더링 해야겠죠.**

2. `ArticleCreatePage` 입니다. 

![image](https://user-images.githubusercontent.com/52649378/122151833-33d85380-ce9b-11eb-970e-f91b4c40d775.png)

위 그림처럼 `boardOptions` 에 그 값을 넣구요.

![image](https://user-images.githubusercontent.com/52649378/122151866-43f03300-ce9b-11eb-851a-1bcd7388fe97.png)

위처럼 `Modal` 창을 만들며 인자로 그 `boardOptions` 를 넘겨줍니다.

3. `ArticleOptionModal` 은 `FilterModal`과 굉장히 비슷합니다. 

다만, 몇가지 설명드릴게요..

![image](https://user-images.githubusercontent.com/52649378/122152021-8ca7ec00-ce9b-11eb-967f-52618264d5ca.png)

- `open`, `close` 는 모달을 띄우고 지우는데 관련된, **모두가 아시리라 믿는** props들입니다.

- 13번째줄 `selectedOptions`, `setSelectedOptions` 는 사용자가 옵션중에 뭐뭐를 클릭했는지를 저장하는 string 배열입니다.

  - 가령 예시를 들자면 `['1분반/월23수3', '질문']` 이런 형식의 데이터가 들어가겠죠.

- 16번째줄은 사용자가 선택한 옵션들을 적용하기 위해 `적용` 버튼을 누를때, `true`값으로 설정하는 일종의 `flag` 입니다.

![image](https://user-images.githubusercontent.com/52649378/122152250-f7592780-ce9b-11eb-9b5e-544940f269fc.png)

- 위 코드는 모든 옵션들 버튼에 바인딩되어 있는 핸들러입니다.

- 위위 사진에 `selectedOptions` 배열에 이미 있다면, 그 값을 삭제하고 없다면 추가하는, 쉽게 말해 **선택한 옵션들에 대해서 토글되는 기능**을 구현했습니다. 

![image](https://user-images.githubusercontent.com/52649378/122152362-21124e80-ce9c-11eb-9018-3c13364ce95a.png)

- 위 함수는 사용자가 `적용` 버튼을 눌렀을때, 실행되는 핸들러입니다.

#### 이렇게 모달 창에 관련된 기능이 구현 설명이 마쳤습니다.

**중간정리를 하자면**,

1. `BoardDetailContainer` 에서 api 콜로 받은 정보를 `atom`에 저장.
2. 이를 바탕으로 `ArticleOptionModal` 에 렌더링.

이렇게 정리할 수 있겠네요.

---

마지막 단계입니다.

4. 이제 선택된 옵션들을 글쓰기 페이지에 노출시켜야합니다. **제목 칸 위에요.**

#### 아래 코드는 `ArticleCreatePage.tsx` 파일 내용입니다.

![image](https://user-images.githubusercontent.com/52649378/122152686-9bdb6980-ce9c-11eb-871a-35d606d04ff3.png)

- 46번째 줄 `selectedOptions` 에는 사용자가 누른 옵션들 string 배열이, 

- 49번째줄은 사용자가 옵션누르고 `적용` 버튼을 눌렀으면 `true` 가 있겠죠.

이 두 정보를 가지고 렌더링합니다.

![image](https://user-images.githubusercontent.com/52649378/122152839-d47b4300-ce9c-11eb-8121-9875095b4658.png)

- 간단하죠? 위위 사진의 49번째줄 변수가 `flag` 이니까 `&&`연산자로 표현했습니다. 

- 137번째줄은 위위 사진의 46번째줄 변수입니다.


---

+) 마지막으로 수정된 사항 알려드립니다.

#### `ArticleCreatePage.tsx` 에서, `options` 관련 부분 수정되었습니다.

![image](https://user-images.githubusercontent.com/52649378/122153012-199f7500-ce9d-11eb-8b21-f96c841f4870.png)

- 71번째 줄이 추가되었습니다. 

  - `newOptions` 변수는 `flag`에 따라 `selectedOptions` or 빈배열값을 가집니다. 

- 이를 88번째 줄에 넣어주며 `options` 값도 게시물 등록할때 무사히 넘기도록 개발했습니다.

---

상세한 코드리뷰 부탁드립니다.

모르는 사항 있으면 물어봐주세요.

Derek-94 committed:
feat(modal): Update article create modal.

- Renders based on recoil value.

Derek-94 committed:
feat(options): Update articleCreateComponent options.

- Have a option modal.
- Toggles when user picks options.
- Remeber prvious selected options, and renders them in red
  background-color.

Derek-94 committed:
feat(articleCreate): Shows options that user picked.

- When user pick options in articleCreate screen, renders the picked
  options top of the title.

Derek-94 committed:
feat(articleCreate): Gives selected options when post new article.

- Gives to container including the options user  selected.

Derek-94 committed:
fix: change string array to empty one.

hyunju95 commented on src/Atoms/atom.tsx:
```diff
@@ -1,4 +1,5 @@
 import { atom } from 'recoil';
+import { OptionResponse } from '../api/ApiProps';
```
from 'api/ApiProps' 로 임포트 해주셔도 될것같습니다~!

Derek-94 replied:
넵 절대경로 깜빡하고 개발이 들어갔네요 수정하겠습니다~

hyunju95 commented on src/Commons/ArticleOptionModal.tsx:
```diff
@@ -0,0 +1,105 @@
+} from 'Atoms/atom';
+
+import 'css/ArticleOptionModal.css';
+import { useRecoilState, useSetRecoilState } from 'recoil';
```
css랑 recoil 임포트 순서 바꿔주세요!

Derek-94 replied:
넵 적용했습니당 :)

hyunju95 commented on src/Commons/ArticleOptionModal.tsx:
```diff
@@ -0,0 +1,105 @@
+  const [selectedOptions, setSelectedOptions] = useRecoilState(
+    articleCreateSelectedOptions
+  );
+  const setApplyArticleCreateOptionsFlag = useSetRecoilState(
```
변수명 관련해서 개발할때마다 드는 생각이

  1. 변수명을 길게 해서 명확하게 용도를 표현할 수 있도록 네이밍한다
  2. 변수명을 짧고 가독성있게 네이밍한다
  
이 두가지 생각으로 매번 고민하는데 
해당 부분의 setApplyArticleCreateOptionsFlag 같은 경우에도 
변수의 용도를 명확하게 표현할 수 있다는 장점이 있지만,
추후에 해당 코드에 대해서 토론을 할 때 해당 변수명을 발음하기도 힘들고 한눈에 알아보기가 쉽지 않은게 단점으로 생각이 되는데 이 부분에 대해서 어떻게 생각하시는지 의견 궁금합니다
다른 좋은방법이 있을지 이야기 나눠보고싶어요!

Derek-94 replied:
넵 한번 회의 해보자구요~

dayonxxi replied:
저도 매번 변수명 작성할때마다 저 두가지 고민을 하게 되는 것 같아요! 그런데 매번 개발할때마다 상황이 다를테니 변수명 작성하는 규칙을 만들기도 어렵고, 만든다해도 유연성 있게 적용은 못하지 않을까 싶네요..!

hyunju95 commented on src/Commons/ArticleOptionModal.tsx:
```diff
@@ -0,0 +1,105 @@
+              </button>
+            ) : (
+              <button
+                className=""
```
혹시 이부분에 className="" 을 아예 없애면 어떻게 동작하나요? 지정할 className이 없는 걸 표현하고 싶은 거라면 아예 className 항목이 없어도 동일한 효과를 가질 것이라고 생각되는데 한번 테스트해보시면 좋을거같아요!

Derek-94 replied:
넵넵 맞네요! 아무래도 없는게 더 나을것 같긴하네요. 수정하겠습니다.

hyunju95 reviewed (commented):
리뷰 남겼습니다~ 개발 수고하셨습니다!!

Derek-94 reviewed (commented):
말씀하신 내용 수정해서 반영했고, 변수명 관련해서는 회의진행해보겠습니다~

Derek-94 committed:
feat(apply): Applied feedback.

- Remove empty className.
- Changed order of imports.
- Changes dir path to absolute.

dayonxxi commented on src/Commons/ArticleOptionModal.tsx:
```diff
@@ -0,0 +1,103 @@
+  const onClickOptions = (e: React.MouseEvent<HTMLButtonElement>) => {
+    const clickedOptionValue = e.currentTarget.value;
+    e.currentTarget.classList.toggle('selected');
+    if (selectedOptions.includes(clickedOptionValue)) {
+      setSelectedOptions(() =>
+        selectedOptions.filter((op) => op !== clickedOptionValue)
+      );
+    } else {
+      setSelectedOptions([...selectedOptions, clickedOptionValue]);
+    }
```
1. e.currentTarget.value로 지금 선택한 option값을 clickedOptionValue에 저장
2. 만약 selectedOptions배열이 clickedOptionValue를 이미 갖고 있다면 filter를 통해 배열에서 빼줌
3. 갖고 있지 않다면 배열에 추가해줌

이렇게 진행되는거 맞을까요?!

Derek-94 replied:
넵 맞습니다~

dayonxxi commented on src/Commons/ArticleOptionModal.tsx:
```diff
@@ -0,0 +1,103 @@
+  };
+
+  return (
+    <div className={open ? 'openArticleOptionModal modal' : 'modal'}>
```
파일 이름이 FilterModal.tsx에서 ArticleOptionModal.tsx로 바꼈네요.
혹시 이부분의 이름은 {open ? 'article-filter-modal-active' : 'article-filter-modal'} 가 어떨까하는 생각이 들었습니다!

Derek-94 replied:
아 그러네요 css는 `-` 포함해서 클래스 네이밍이 들어가야하는데.. 수정하겠습니다!

dayonxxi reviewed (commented):
리뷰 남겼습니다! 수고하셨어요 :)

hyunju95 reviewed (approved):
수정하신부분 확인했습니다 수고하셨어요!!

Derek-94 committed:
style: rename className.

- Renaming className.

Derek-94 reviewed (commented):
수정했습니다!

Derek-94 committed:
Merge branch 'develop' of https://github.com/dev-floor/untitled-client into feature/54
